src/db.c

```diff
@@ -1714,7 +1714,22 @@ int expireIfNeeded(redisDb *db, robj *key, int flags) {
 
     /* In some cases we're explicitly instructed to return an indication of a
      * missing key without actually deleting it, even on masters. */
-    if (flags & EXPIRE_AVOID_DELETE_EXPIRED)
+    int avoid_del = flags & EXPIRE_AVOID_DELETE_EXPIRED;
+
+    /* If we are running a command that may modify random keys, from within a
+     * parent execution unit (i.e. it must be wrapped in MULTI/EXEC) we must
+     * avoid lazy-expire when in cluster mode (otherwise the replica may
+     * return a CROSSSLOT error for the transaction) */
+    if (server.cluster_enabled &&
+        server.execution_nesting > 1 &&
+        server.current_client &&
+        server.current_client->cmd &&
+        server.current_client->cmd->flags & CMD_MODIFY_RANDOM_KEYS)
+    {
+        avoid_del = 1;
+    }
+
+    if (avoid_del)
         return 1;
```

Comment:
@guybe7 you mentioned to me you found a bug here, that we return 1 instead of 0, letting the caller know the key exists, when all we wanted to do is avoid the deletion.
is that right or are we missing something?

currently only used in handleClientsBlockedOnKey and getNodeByQuery...

Reply:
ohh, sorry..  i now realize returning 1 means the key is missing, so there's no bug here.
and in fact what you were concerned about was `lazy_expire_disabled` always returning 0 (as if they key isn't logically expired).

i see this flag was added for moduleNotifyKeyUnlink where we actually want to let the module access the key.
what else am i missing?

Reply:
`moduleNotifyKeyUnlink` passes the value to both the `unlink` CB and `moduleFireServerEvent`
so the module gets the value regardless of `lazy_expire_disabled`

and `lazy_expire_disabled` is also used in `moduleNotifyKeyspaceEvent`

i believe the purpose was to prevent innocent reeantrancy: module tries to access a random key from within the event (or notify CB, or unlink) and is triggered again because the access caused a lazy-expire causing deletion/notification

if that's indeed the case, we should only avoid the actual deletion (and notification obviously) of the key, but lookupKey should still return NULL

Reply:
initially, moduleNotifyKeyUnlink didn't hand the value to the module, it would just hand it the key name, and we expected the module to open the key, or even do RM_Call to interrogate it (some details don't have a module API to access them). so i don't think it's only about avoiding deletion.
regarding moduleNotifyKeyspaceEvent i'm not certain, maybe we did just want to avoid reeantrancy.
considering this is only modules (which are power users. previously i thought we're talking about handleClientsBlockedOnKey and getNodeByQuery), i think i'm ok leaving it as is.